napari/layers/_shapes_layer/model.py

```diff
@@ -1571,86 +1514,153 @@ def on_mouse_release(self, event):
         else:
             raise ValueError("Mode not recongnized")
 
-    def on_key_press(self, event):
-        """Called whenever key pressed in canvas.
+    def _activate_add_rectangle_mode(self):
+        self.mode = Mode.ADD_RECTANGLE
 
-        Parameters
-        ----------
-        event : Event
-            Vispy event
+    def _activate_add_elipse_mode(self):
+        self.mode = Mode.ADD_ELLIPSE
+
+    def _activate_add_line_mode(self):
+        self.mode = Mode.ADD_LINE
+
+    def _activate_add_path_mode(self):
+        self.mode = Mode.ADD_PATH
+
+    def _activate_add_polygon_mode(self):
+        self.mode = Mode.ADD_POLYGON
+
+    def _activate_direct_mode(self):
+        self.mode = Mode.DIRECT
+
+    def _activate_select_mode(self):
+        self.mode = Mode.SELECT
+
+    def _activate_pan_zoom_mode(self):
+        self.mode = Mode.PAN_ZOOM
+
+    def _activate_vertex_insert_mode(self):
+        self.mode = Mode.VERTEX_INSERT
```

Comment:
these all feel like they should be lambdas...?

Reply:
I was considering this but I think in the future we may want to have access to these functions in for example a YAML file (via some parsing):
```yaml
Shapes:
  Control-i: activate_vertex_insert_mode
```

Reply:
Side note: I'm not sure about separate sections in the yaml for each layer type, but in any case I would want it to include the full path to that class:

```yaml
napari.layers.Shapes:
    Ctrl-i: activate_vertex_insert_mode
```

calls `napari.layers.Shapes.activate_vertex_inside_mode`. (Note: the use of underscores should be consistent — which is a conundrum.)

I think we need to think long and hard about what callbacks accessible from the key config should get as input, *and this should always be consistent*. Is it the topmost layer? Is it the list of selected layers? Is it the Layers object itself? Is it the whole viewer?

I think it's ok for the bindings to change depending on the state of the viewer, but do we want the *inputs* to change depending on the state of the viewer? That could make it really difficult to keep everything straight as a plugin developer.

If we do define different callback inputs for different states, then we need to define what those states *can* be, and this should be a finite list.

I guess what you're going for here is that each layer can define its own shortcuts, and these shortcuts would be enabled in the case *and only in the case* where that is the only selected layer. In that case, the config file could look like:

```yaml
- base:
  - [base keybindings for all of napari]
- single-selection:
  - napari.layers.Shapes:
    - ctrl-i: path.to.my_function
- multi-selection:
  - crtl-up: napari.layers.shift_selected_layers_up
```

In that case, each top-level section could have different input arguments to the callbacks.

I really like the idea of full path specs on the functions, because then adding all kinds of crazy functionality could be added with a pip install and a .yaml file.